adapters/tlsio_mbedtls.c

```diff
@@ -614,34 +641,28 @@ int tlsio_mbedtls_send(CONCRETE_IO_HANDLE tls_io, const void* buffer, size_t siz
             }
         }
     }
+
     return result;
 }
 
 void tlsio_mbedtls_dowork(CONCRETE_IO_HANDLE tls_io)
 {
     if (tls_io != NULL)
     {
-        TLS_IO_INSTANCE* tls_io_instance = (TLS_IO_INSTANCE*)tls_io;
-
-        if ((tls_io_instance->tlsio_state != TLSIO_STATE_NOT_OPEN) &&
-            (tls_io_instance->tlsio_state != TLSIO_STATE_ERROR))
+        TLS_IO_INSTANCE *tls_io_instance = (TLS_IO_INSTANCE *)tls_io;
+        if (tls_io_instance->tlsio_state == TLSIO_STATE_OPENING_UNDERLYING_IO || tls_io_instance->tlsio_state == TLSIO_STATE_IN_HANDSHAKE || tls_io_instance->tlsio_state == TLSIO_STATE_OPEN)
         {
             decode_ssl_received_bytes(tls_io_instance);
-            xio_dowork(tls_io_instance->socket_io);
+            // No need to call xio_dowork here because it's called in on_io_recv which is the callback function of decode_ssl_received_bytes
```

Comment:
>xio_dowork [](start = 31, length = 10)

If there is no harm in calling xio_dowork, let us continue calling it here too. Please add a TODO comment to assess this change in the future.

Reply:
perf downgrade and make debug harder